### PR TITLE
Beydoun/spa optics 2

### PR DESCRIPTION
**Pull request by hassanbeydoun**
This PR adds the optics portion of simple-prescribed-aerosol (SPA). 3d_4d and 4d_4d infld options were added to ncdio_atm.F90 since aerosol optical properties are 3d (X,Z, and radiation bands) variables. SPA optics reads in 1 lw aerosol optical property and 3 sw aerosol optical properties.

### CODE REVIEW

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [inline components/eam/src/physics/rrtmgp/radiation.F90:116]: Just to make sure I'm not missing something: you're initializing to false here but then overwriting with whatever was specified in the namelist, right?

Comment (hassanbeydoun) [reply components/eam/src/physics/rrtmgp/radiation.F90:116]: That's right.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [inline components/eam/src/physics/rrtmgp/radiation.F90:470]: This is ok for right now, but is hardcoded for LC machines. We will need to move this file to inputdata and add it to the namelist specification (perhaps in another PR)

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [inline components/eam/src/physics/rrtmgp/radiation.F90:478]: these writeouts are probably not needed once debugging is over. If they are permanent, you should be more clear about what they are. E.g. "file name" => "Prescribed aerosol file name"...

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (AaronDonahue) [reply components/eam/src/physics/rrtmgp/radiation.F90:470]: It might be good to change this in the namelist now.  No one but Hassan will have access to this folder unless he changes the permissions.  @hassanbeydoun , you can follow the example of any of the input file variables in the namelist_definition and namelist_defaults files.  For example, the most recent would be the 'p3_lookup' tables.  You will also need to copy the aerosol optics files to the usr/climdat/... folder so that it is accessible by all.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (AaronDonahue) [inline components/eam/src/physics/rrtmgp/radiation.F90:982]: The use of allocate and deallocate in this chunk of code I think is a way to re-initialize and recycle the `aerosol_optical_property` array.  I think you can also just get away with setting the whole array to a value each time, like `aerosol_optical_property(:,:,:,:) = 0.0` or something along those lines.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (AaronDonahue) [inline components/eam/src/physics/rrtmgp/radiation.F90:991]: Of course if you decide to remove some of the deallocation steps you will need to make sure you keep the one that happens before you switch size from `lwbands` to `swbands`.

Alternatively, you can just declare `aerosol_optical_property_lw` and `aerosol_optical_property_sw` arrays at the top and not have to `allocate/deallocate` anything.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (AaronDonahue) [inline components/eam/src/physics/rrtmgp/radiation.F90:1345]: Maybe add a comment here that SPA optics is not compatible with Leap Year.  It goes without saying really, but if in sometime in the distant future someone does decide to add leap year in it's nice to give them a fighting chance to find all the changes they need to make.  Being able to grep "leap year" might be something that they do.

**Review by AaronDonahue — CHANGES_REQUESTED** [review]
Looks good Hassan! I had a few comments and a followup on Peter's comment about the hard-coded path to the input files.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [reply components/eam/src/physics/rrtmgp/radiation.F90:470]: Agreed. It is easy to do and won't be useable by anyone other than Hassan if he doesn't do this.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [reply components/eam/src/physics/rrtmgp/radiation.F90:991]: Good idea, Aaron. I like the idea of just allocating both aerosol_optical_property_lw and aerosol_optical_property_sw at the top of the subroutine and just assigning new values for each property. I'm not sure whether zeroing out in between usage is needed - it is safer, but if `get_aerosol_optical_property_from_file` is guaranteed to produce a slab of data of the same size as the allocated chunk of memory (which I think is the case), there's no point and skipping the zeroing would be faster and cleaner.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [reply components/eam/src/physics/rrtmgp/radiation.F90:1345]: Good point. It may be worth adding the same comment in the microphysics SPA code as well.

**Comment by PeterCaldwell** [comment]
Looks great, @hassanbeydoun ! I particularly like how `get_aerosol_optical_property_from_file` is reused to keep the SPA sections compact and easy to understand. I think there are just a couple easy fixes left:

1. replace hard-coded file location with standard-practice location. 
2. delete debugging print statements
3. only allocate/deallocate 1x
4. add comment about noleap calendar

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [inline components/eam/src/physics/rrtmgp/radiation.F90:479]: Might as well delete this commented out line.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [inline components/eam/src/physics/rrtmgp/radiation.F90:1000]: Do you need deallocate statements here?

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [inline components/eam/src/physics/rrtmgp/radiation.F90:1644]: Again, do you need to deallocate here?

File: run_scream.syrah_ne30_standard.csh
Comment (PeterCaldwell) [inline run_scream.syrah_ne30_standard.csh:1]: I don't think you meant to include this file in the PR.

Comment (PeterCaldwell) [reply run_scream.syrah_ne30_standard.csh:1]: Also for the other run_scream file you added too.

**Comment by PeterCaldwell** [comment]
@hassanbeydoun - I just made some trivial comments to address. Once you fix them, please ping Aaron and Ben for a re-review. I'd also like to ask what runs you've done to test this capability. 

I notice that autotester is failing for this PR. I expected this because you've changed the variables being passed to/from functions so they're inconsistent with the C++. So we need to make the C++ version of these changes before we can merge to master. We could potentially merge to dyamond2 branch without making those changes, but I'd rather more thoroughly test the SPA implementation in this branch before doing any merging. I'll think about/try to find someone to add the C++ side.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (hassanbeydoun) [reply components/eam/src/physics/rrtmgp/radiation.F90]: deleted. Missed it during last cleanup.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (hassanbeydoun) [reply components/eam/src/physics/rrtmgp/radiation.F90:996]: it worked without deallocation but I added it just in case.

File: run_scream.syrah_ne30_standard.csh
Comment (hassanbeydoun) [reply run_scream.syrah_ne30_standard.csh]: Deleted. thought I deleted it last time.

**Comment by hassanbeydoun** [comment]
> @hassanbeydoun - I just made some trivial comments to address. Once you fix them, please ping Aaron and Ben for a re-review. I'd also like to ask what runs you've done to test this capability.
> 
> I notice that autotester is failing for this PR. I expected this because you've changed the variables being passed to/from functions so they're inconsistent with the C++. So we need to make the C++ version of these changes before we can merge to master. We could potentially merge to dyamond2 branch without making those changes, but I'd rather more thoroughly test the SPA implementation in this branch before doing any merging. I'll think about/try to find someone to add the C++ side.

**Review by AaronDonahue — APPROVED** [review]
Looks good, approving assuming the you have checked that the tests pass.

File: components/eam/src/physics/rrtmgp/radiation.F90
Comment (PeterCaldwell) [reply components/eam/src/physics/rrtmgp/radiation.F90:996]: I think this kind of stuff causes memory leaks. Maybe not in this case since this is just in init, but probably in the radiation_tend case where you allocate every month but never deallocate. In any case, fixed now.

**Review by PeterCaldwell — APPROVED** [review]

**Comment by PeterCaldwell** [comment]
@brhillman - could you review?
@hassanbeydoun - it looks like the AT fail is happening to everyone so I'll manually merge once Ben reviews and I hear from you that sims with and without SPA look vaguely similar.

**Comment by PeterCaldwell** [comment]
I just confirmed that default (NON-SPA) behavior passes ERP and SMS_D tests at ne4pg2 on cori and syrah. ERP flagged a memory leak on Cori which @ndkeen tracked down to the memory leak we had on master when this branch was created. Adding the relevant bugfix (not included in this PR) fixed the problem. Thanks Noel! 

I have not been able to test do_spa_optics =.true. because default prescribed optics files are not available yet. We will add them in a future PR.

**Pull request merged**
